lib/at_parser/at_parser.c

```diff
@@ -0,0 +1,548 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <string.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <errno.h>
+#include <zephyr/sys/util.h>
+#include <modem/at_parser.h>
+
+#include "at_token.h"
+#include "at_match.h"
+
+/* Carriage Return. */
+#define CR '\r'
+/* Line Feed. */
+#define LF '\n'
+/* Null Terminator. */
+#define NULL_TERMINATOR '\0'
+/* Minus Sign. */
+#define MINUS_SIGN '-'
+/* Init Sentinel. */
+#define INIT_SENTINEL 0xdeadbeef
+
+enum at_num_type {
+	AT_NUM_TYPE_INT16,
+	AT_NUM_TYPE_UINT16,
+	AT_NUM_TYPE_INT32,
+	AT_NUM_TYPE_UINT32,
+	AT_NUM_TYPE_INT64,
+	AT_NUM_TYPE_UINT64,
+};
+
+/* Trim expected CR, LF, or CRLF. */
+static void trim_crlf(const char **str)
+{
+	while ((*str)[0] == CR || (*str)[0] == LF) {
+		(*str)++;
+	}
+}
+
+static bool is_subparam(const struct at_token *token)
+{
+	switch (token->type) {
+	case AT_TOKEN_TYPE_INT:
+	case AT_TOKEN_TYPE_QUOTED_STRING:
+	case AT_TOKEN_TYPE_ARRAY:
+	case AT_TOKEN_TYPE_EMPTY:
+		return true;
+	default:
+		return false;
+	}
+}
+
+static bool is_resp(const char *str)
+{
+	/* All possible response variants. */
+	static const char * const resp[] = {
+		"OK\r\n",
+		"ERROR\r\n",
+		"+CME ERROR",
+		"+CMS ERROR"
+	};
+
+	trim_crlf(&str);
+
+	for (size_t i = 0; i < ARRAY_SIZE(resp); i++) {
+		if (strncmp(str, resp[i], strlen(resp[i])) == 0) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+static bool is_index_ahead(struct at_parser *parser, size_t index)
+{
+	return index + 1 > parser->count;
```

Comment:
To me it is clearer this way because of what index and count mean. For example, the first element has index 0, but count 1. The inequality with the `+ 1`  normalizes this comparison and the `>` shows that the element is ahead.